File: multitone.lua
--- Module to generate a duotone effect.
-- It also supports gradients of more than two colors.
local base = [[
<filter id='$id'>
	<feColorMatrix type="matrix" values=".375 .5 .125 0 0 .375 .5 .125 0 0 .375 .5 .125 0 0 0 0 0 1 0" />
	<feComponentTransfer color-interpolation-filter="sRGB">
		<feFuncR type="table" tableValues="$red" />
		<feFuncG type="table" tableValues="$green" />
		<feFuncB type="table" tableValues="$blue" />
	</feComponentTransfer>
</filter>
]]

local function normal(number)
	return math.min(1, math.max(0, math.floor(number*1000)/1000))
end

local function multitone(id, ...)
	local components = {{}, {}, {}}
	for i, color in ipairs({...}) do
		components[1][i] = tostring(normal(color[1]))
		components[2][i] = tostring(normal(color[2]))
		components[3][i] = tostring(normal(color[3]))
	end
	local options = {
		red = table.concat(components[1], ' ');
		green = table.concat(components[2], ' ');
		blue = table.concat(components[3], ' ');
		id = id
	}
	return (base:gsub("$%w+", function(key)
		return options[key:sub(2)] or error("Key "..key.." not found in options!")
	end))
end

return multitone

File: multitone.js
const normal = number => Math.min(1, Math.max(0, Math.floor(number*1000)/1000))

const element = (name, options, ...children) => {
	const result = document.createElementNS('http://www.w3.org/2000/svg', name)
	Object.entries(options).forEach(([key, value]) => {
		result.setAttribute(key, value)
	})
	result.append(...children)
	return result
}

export default (id, ...colors) => {
	const components = [[], [], []]
	colors.forEach(color => {
		for (let i=0; i<3; i++) {
			components[i].push(normal(color[i]).toString())
		}
	})
	return element('filter', {id},
		element("feColorMatrix", {type: 'matrix', values: '.375 .5 .125 0 0 .375 .5 .125 0 0 .375 .5 .125 0 0 0 0 0 1 0'}),
		element("feComponentTransfer", {"color-interpolation-filter": "sRGB"},
			element("feFuncR", {type: "table", tableValues: components[0].join(" ")}),
			element("feFuncG", {type: "table", tableValues: components[1].join(" ")}),
			element("feFuncB", {type: "table", tableValues: components[2].join(" ")})
		)
	)
}
